src/appchooserdialog.c

```diff
@@ -60,56 +74,248 @@ app_chooser_dialog_init (AppChooserDialog *dialog)
+  g_object_set (message_dialog,
+                "text", primary,
+                "secondary-text", secondary,
+                NULL);
+  gtk_dialog_set_default_response (GTK_DIALOG (message_dialog), GTK_RESPONSE_OK);
+  gtk_widget_show (message_dialog);
+  g_signal_connect (message_dialog, "response",
+                    G_CALLBACK (gtk_widget_destroy), NULL);
 }
 
 static void
-button_clicked (GtkWidget *button,
+link_activated (GtkWidget *label,
+                const char *uri,
                 AppChooserDialog *dialog)
 {
-  gtk_widget_hide (GTK_WIDGET (dialog));
-  g_signal_emit (dialog, signals[DONE], 0, NULL);
+  g_autofree char *option = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  g_autoptr(GError) error = NULL;
+
+  if (dialog->content_type)
+    option = g_strconcat ("--search=", dialog->content_type, NULL);
+  else
+    option = g_strdup ("--mode=overview");
+
+  process = g_subprocess_new (0, &error, "gnome-software", option, NULL);
+  if (!process)
+    show_error_dialog (_("Failed to start Software"), error->message, GTK_WINDOW (dialog));
+}
+
+static void
+populate_full_list (AppChooserDialog *dialog)
+{
+  GList *apps, *l;
+
+  apps = g_app_info_get_all ();
+
+  for (l = apps; l; l = l->next)
+    {
+      GAppInfo *info = l->data;
+      GtkWidget *row;
+
+      row = GTK_WIDGET (app_chooser_row_new (info));
+      gtk_widget_set_visible (row, TRUE);
+      gtk_flow_box_insert (GTK_FLOW_BOX (dialog->full_list), row, -1);
+    }
+
+  g_list_free_full (apps, g_object_unref);
+}
+
+static gboolean
+scroll_down (gpointer data)
+{
+  AppChooserDialog *dialog = data;
+  GtkAdjustment *adj;
+  GtkAllocation alloc;
+
+  gtk_widget_get_allocation (dialog->full_list, &alloc);
+  adj = gtk_scrolled_window_get_vadjustment (GTK_SCROLLED_WINDOW (dialog->scrolled_window));
+  gtk_adjustment_set_value (adj, alloc.y);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+show_full_list (AppChooserDialog *dialog)
+{
+  if (!gtk_widget_get_visible (dialog->more_button))
+    return;
+
+  gtk_scrolled_window_set_policy (GTK_SCROLLED_WINDOW (dialog->scrolled_window),
+                                  GTK_POLICY_NEVER,
+                                  GTK_POLICY_AUTOMATIC);
+
+  gtk_widget_hide (dialog->more_button);
+  gtk_widget_show (dialog->search_button);
+  gtk_widget_show (dialog->full_list_box);
+
+  populate_full_list (dialog);
+
+  g_idle_add (scroll_down, dialog);
+}
+
+static void
+more_clicked (GtkButton *button,
+              AppChooserDialog *dialog)
+{
+  show_full_list (dialog);
+}
+
+static gboolean
+filter_func (GtkFlowBoxChild *child,
+             gpointer data)
+{
+  AppChooserRow *row = APP_CHOOSER_ROW (child);
+  AppChooserDialog *dialog = data;
+  GAppInfo *info;
+  char *name;
+
+  if (!dialog->search_text)
+    return TRUE;
+
+  info = app_chooser_row_get_info (row);
+
+  name = g_utf8_casefold (g_app_info_get_name (info), -1);
+  if (strncmp (name, dialog->search_text, strlen (dialog->search_text)) == 0)
```

Comment:
Fixed. There was also a memory leak hiding in plain sight here. Fixed that as well.